releases/release-4.0.12.md

```diff
@@ -0,0 +1,117 @@
+---
+title: TiDB 4.0.12 Release Notes
+---
+
+# TiDB 4.0.12 Release Notes
+
+Release date: March 31, 2021
+
+TiDB version: 4.0.12
+
+## New Features
+
++ TiFlash
+
+    - Add tools to check exact tiflash replica status for online rolling update [#1512](https://github.com/pingcap/tics/pull/1512)
+
+## Improvements
+
++ TiFlash
+
+    - Automatically clean archive data to free up disk space [#1638](https://github.com/pingcap/tics/pull/1638)
+    - Allow omitting deprecated "users.*" settings in the configuration file of TiFlash. [#1535](https://github.com/pingcap/tics/pull/1535)
+    - Reduce the size of TiFlash binary file. [#1414](https://github.com/pingcap/tics/pull/1414)
+
+## Bug Fixes
+
++ TiDB
+
+    - Fix get var expr when session var is hex literal [#23372](https://github.com/pingcap/tidb/pull/23372)
+    - Fix the bug that wrong collation is used when try fast path for enum or set [#23292](https://github.com/pingcap/tidb/pull/23292)
+    - Wrong result of nullif expr when used with is null expr [#23279](https://github.com/pingcap/tidb/pull/23279)
+    - Statistics: fix a case that auto-analyze is triggered outside its time range [#23219](https://github.com/pingcap/tidb/pull/23219)
+    - Executor: fix cast function will ignore the error for point get key construction [#23211](https://github.com/pingcap/tidb/pull/23211)
+    - Fixed a bug that prevented SPM from taking effect [#23209](https://github.com/pingcap/tidb/pull/23209)
+    - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
+    - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
+    - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
+    - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
+    - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
+    - Scattering truncated tables without pre-split option. [#22872](https://github.com/pingcap/tidb/pull/22872)
+    - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
+    - Add three format specifier for str_to_date expression [#22812](https://github.com/pingcap/tidb/pull/22812)
+    - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
+    - Record prepare execute fail as "Failed Query OPM" in monitor [#22672](https://github.com/pingcap/tidb/pull/22672)
+    - Do not report error for prepared stmt execution if tidb_snapshot is set [#22641](https://github.com/pingcap/tidb/pull/22641)
```

Comment:
```suggestion
    - Fix get var expr when session var is hex literal [#23372](https://github.com/pingcap/tidb/pull/23372)
    - Fix the bug that wrong collation is used when try fast path for enum or set [#23292](https://github.com/pingcap/tidb/pull/23292)
    - Wrong result of nullif expr when used with is null expr [#23279](https://github.com/pingcap/tidb/pull/23279)
    - Statistics: fix a case that auto-analyze is triggered outside its time range [#23219](https://github.com/pingcap/tidb/pull/23219)
    - Executor: fix cast function will ignore the error for point get key construction [#23211](https://github.com/pingcap/tidb/pull/23211)
    - Fixed a bug that prevented SPM from taking effect [#23209](https://github.com/pingcap/tidb/pull/23209)
    - Fix wrong table filters for index merge plan [#23165](https://github.com/pingcap/tidb/pull/23165)
    - Fix unexpected NotNullFlag in case when expr ret type [#23135](https://github.com/pingcap/tidb/pull/23135)
    - Fix a bug that collation is not handle for text type [#23092](https://github.com/pingcap/tidb/pull/23092)
    - Fix range partition prune bug for IN expr [#23074](https://github.com/pingcap/tidb/pull/23074)
    - Fix request block forever when replace a tombstone store with the store has same IP addr [#23071](https://github.com/pingcap/tidb/pull/23071)
    - Do not ajust int when it is null and compared year [#22844](https://github.com/pingcap/tidb/pull/22844)
    - Fix load data lost connection error on tables with auto_random column [#22736](https://github.com/pingcap/tidb/pull/22736)
    - Fix ddl hang over when it meets panic in cancelling path. [#23297](https://github.com/pingcap/tidb/pull/23297)
    - Fix wrong key range of index scan when filter is comparing year column with NULL [#23104](https://github.com/pingcap/tidb/pull/23104)
    - Fix create view success but failed when using it [#23083](https://github.com/pingcap/tidb/pull/23083)
```